src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -0,0 +1,285 @@
+        // note that 1/1/1601 will be returned as a negative value by Java
+        GregorianCalendar today = new GregorianCalendar();
+        Date todayDate = today.getTime();
+        long timeSinceWin32EpochInMs = todayDate.getTime() - win32EpochDate.getTime();
+        // milliseconds to microseconds => x1000
+        long timeSinceWin32EpochInNs = TimeUnit.NANOSECONDS.convert(timeSinceWin32EpochInMs, TimeUnit.MILLISECONDS);
+        // but we need in 100 ns, as 1000 ns = 1 micro, add a x10 factor
+        return timeSinceWin32EpochInNs * 100;
+    }
+
+    // https://ldapwiki.com/wiki/Password%20Expiration
+    public static boolean checkIfCredentialsAreNonExpired(@Nonnull Attributes user) {
+        // Active Directory attributes
+        Integer uac = getUserAccountControl(user);
+        if (uac != null) {
+            if ((uac & ADS_DONT_EXPIRE_PASSWORD) == ADS_DONT_EXPIRE_PASSWORD) {
+                return true;
+            }
+            if ((uac & ADS_UF_PASSWORD_EXPIRED) == ADS_UF_PASSWORD_EXPIRED) {
+                return false;
+            }
+        }
+        Boolean passwordExpired = getBooleanAttribute(user, ATTR_USER_PASSWORD_EXPIRED);
+        if (passwordExpired != null) {
+            return !passwordExpired;
+        }
+        // no other indicators
+        return true;
+    }
+
+    // https://ldapwiki.com/wiki/Account%20Lockout
+    // https://ldapwiki.com/wiki/Intruder%20Detection
+    public static boolean checkIfAccountNonLocked(@Nonnull Attributes user) {
+        // Active Directory attributes
+        Integer uac = getUserAccountControl(user);
+        if (uac != null && (uac & ADS_UF_LOCK_OUT) == ADS_UF_LOCK_OUT) {
+            return false;
+        }
+        // standard attributes
+        Boolean lockout = getBooleanAttribute(user, ATTR_PWD_LOCKOUT);
+        if (lockout != null) {
+            return !lockout;
+        }
+        // EDirectory attributes
+        Boolean lockedByIntruder = getBooleanAttribute(user, ATTR_LOCKED_BY_INTRUDER);
+        if (lockedByIntruder != null) {
+            return !lockedByIntruder;
+        }
+        // no other indicators
+        return true;
+    }
+
+    private static @CheckForNull Integer getUserAccountControl(@Nonnull Attributes user) {
+        String uac = getStringAttribute(user, ATTR_USER_ACCOUNT_CONTROL);
+        String computedUac = getStringAttribute(user, ATTR_USER_ACCOUNT_CONTROL_COMPUTED);
+        if (uac == null) {
+            return computedUac == null ? null : Integer.parseInt(computedUac);
+        } else if (computedUac == null) {
+            return Integer.parseInt(uac);
+        } else {
+            return Integer.parseInt(uac) | Integer.parseInt(computedUac);
+        }
+    }
+
+    private static @CheckForNull Boolean getBooleanAttribute(@Nonnull Attributes user, @Nonnull String attrID) {
+        String attr = getStringAttribute(user, attrID);
+        return attr == null ? null : "true".equalsIgnoreCase(attr);
```

Comment:
Any reason not to use Boolean.valueOf(String) ?

Reply:
Doesn't support `null`.

Reply:
(This is really a nit, but since we started discussing it, I'll continue a bit. It's not a blocker.)
The documentation indicates otherwise.

```
public static Boolean valueOf(String s)

Returns a Boolean with a value represented by the specified string. The Boolean returned represents a true value if the string argument is not null and is equal, ignoring case, to the string "true".
```

I guess yours results in a tri-valued Boolean (true, false, or null), whereas I suspect the standard one only returns (Boolean.FALSE, Boolean.TRUE). If you need that third value for some reason, then yours is better, otherwise it would be better to have a dual-valued boolean.